[FEATURE] Ability to clear all states for a given state name

offirgolan opened the pull request:
Use Cases:

- Easily teardown an app's state without having to teardown the actual app
- Easily clear state between test cases

stefanpenner commented:
> Easily teardown an app's state without having to teardown the actual app
> Easily clear state between test cases

How does this work with the weakmap polyfil?

offirgolan commented:
@stefanpenner this doesnt touch the weakmaps. Just sets the given stateName property on the  `weakMaps` pojo to null. That way when the cp looks for a state, it would have to create a new weakmap and therefore a new state.

thoov commented:
I think clearing state is something that we need. My initial reaction for adding something like `clearStates()` is that we might be fighting a design pattern (meaning that we might have to look at a state manager). However that is a bigger / tougher problem than just this use case. I think that in the short term we can add this (pending that it can solve the case were clearing a state with active CPs will clear them and not point to old / stale data +  some test coverage) and we can start a discussion around a potential adding a different delivery method that stateFor.

stefanpenner commented:
> @stefanpenner this doesnt touch the weakmaps.

so we leak those resources? Something seems fishy about that.

offirgolan commented:
@stefanpenner I'll modify this PR with an updated solution and some test cases. 

Here is an example use case:

A user belongs to multiple teams. Switching teams should reset all the state across the application to have a clean start. To do this currently is really difficult since you have to manage every single component state. 


Another use case is when you want to clear state between acceptance tests. 

In regards to leaking resources, that shouldnt be a problem if the CP recomputes when clearing the states since the state weakmaps should be completely detached at that point.

stefanpenner commented:
> Another use case is when you want to clear state between acceptance tests.

The data stored via weak-maps actually lives on the entities (weakMap keys) retained by the container, which means between acceptance tests no state should survive.

> A user belongs to multiple teams. Switching teams should reset all the state across the application to have a clean start. To do this currently is really difficult since you have to manage every single component state.

If all state is being reset, then the weak-maps wont reference anything as they are stateless.

What am I missing that makes my above statements not correct or not adequate.

offirgolan commented:
> If all state is being reset, then the weak-maps wont reference anything as they are stateless.

The app still keeps state while you use it. Only once switching to a different team, all state has to be reset / cleaned up.

stefanpenner commented:
> Only once switching to a different team, all state has to be reset / cleaned up.

it isn't clear to me what you mean.

offirgolan commented:
Okay let me try to break this down :) 

1. User logs into **Team 1** in the app
2. User navigates around the app, searches data tables, etc. which all use this addon to maintain state.
3. User then decides to switch to **Team 2**
4. At the point all state that was created for **Team 1** should now be cleared so that the user has a fresh start with the new team. 
5. User navigates around the app, searches data tables, etc. which all use this addon to maintain state.

stefanpenner commented:
Ok, thank you.

The following is the point I am confused about:

> At the point all state that was created for Team 1 should now be cleared so that the user has a fresh start with the new team.

If all state is removed, what state remains that the weak-map gives access to?

offirgolan commented:
> If all state is removed, what state remains that the weak-map gives access to?

Ah okay. So we have files in the app that are bound to the state for the life-time of the app. 

```js
// in app/tables/my-table.js
export default new Table()
```

Then in a component or w/e

```js
import myTable from '../tables/my-table';

export default Ember.Component.extend({
  table: myTable,
  state: stateFor('tables', 'table')
});
```

So when the component gets destroyed the state is still alive for the table until the app is torn down. 

Right now we can do something like

```js
let stateMap = stateFor('tables');

if(stateMap && stateMap.has(myTable)) {
  stateMap.delete(myTable);
}
```

That solution works find BUT we have different tables across the entire app. So to do this, we have to import every table file and do the above step which isnt very convenient nor scalable. It would be super nice to do something like `clearState('tables')` instead.

stefanpenner commented:
> So we have files in the app that are bound to the state for the life-time of the app.

This sounds like the root problem, but in general we aim to keep state managed by the container. Anything at the module level is to be stateless. Any deviation from this really feels like it creates more problems then it solves. Please share why this state can't be managed by the container?

offirgolan commented:
> Please share why mutation global state (e.g. the modules) is required

Here is one of the use cases that we currently have:

1. User goes to table listing page
2. User filters / sorts / searches the list 
3. User scrolls down the list
4. User selects a row and enters that record's detail route
5. User then clicks Save / Done / Back 
6. User should be back at the listing page with the same scroll position, same data, same filter, etc. 

Since all logic is contained to different components which get destroyed when switching routes, for us to keep state across the multiple routes, **something** has to be constant for this addon to maintain the correct state object. That something is the global table modules.

stefanpenner commented:
> That something is the global table modules.

This path is fraught with peril. Other memory leaks are likely to happen, as it contamination between sessions, which is what you are attempting to work around here.

My recommendation is to put this state elsewhere, ideally something pinned to the container. At which point, everything will "just work" and no risk of memory leaks will be present.

offirgolan commented:
@stefanpenner I came up with an alternate solution on my end. 

1. I created a simple "event bus" service via Ember.Evented that triggers a "team switched" event
2. I moved the table instance from a global module to it's appropriate controller and passed it to the components
3. I added a listener on that controller so when the user switches teams, I just create a new table instance which then creates a new state bucket.

That works quiet well and I'm able to contain state cleanup to the appropriate controllers, however, I feel like there should be some sort of way in the future to easily cleanup unwanted states?

stefanpenner commented:
> I created a simple "event bus" service via Ember.Evented that triggers a "team switched" event
> I moved the table instance from a global module to it's appropriate controller and passed it to the components
> I added a listener on that controller so when the user switches teams, I just create a new table instance which then creates a new state bucket.

Another approach is attaching this table state to w/e knows about the currentTeam. That way, when that changes, it (without needing any complexity) could release the table.

```js
// services/team.js

export default Ember.Service({
  init() {
    this._super(...arguments);
    this.currentTeam = ...
  },

  table: computed('currentTeam', function() {
    // create a new instance of some kind with table state.
  });
});
```

Now when you change `currentTeam` on the *team service* the worry-some state is automatically released, as is it when the container is destroyed.

stefanpenner commented:
> That works quiet well and I'm able to contain state cleanup to the appropriate controllers, however, I feel like there should be some sort of way in the future to easily cleanup unwanted states?

Maybe, but in nearly all scenarios I have ever seen, doing so actually hides other issues. (As we described above)

offirgolan commented:
> The owner of state, is the one who should destroy it

I think that's reasonable. My alternate solution actually solved all of my problems 😛. I'll go ahead and close this PR then. If you want me to pursue a better solution in the future, feel free to let me know.
